[BC Break] React Hook Form (#7087)
marmelab/react-admin: UPGRADE.md

}
```

## react-final-form Has Been Replaced By react-hook-form

### `<FormWithRedirect>` Has Been Renamed to `<Form>`

The form components don't handle redirection at all anymore as it happens in side effects, the `<FormWithRedirect>` has been renamed to `<Form>`.

> To upgrade simply replace all occurrences of `FormWithRedirect` with  `Form`.

Reviewer: ```suggestion
To upgrade, replace all occurrences of `<FormWithRedirect>` with  `<Form>`.
```

```diff
- import { FormWithRedirect } from 'react-admin';
+ import { Form } from 'react-admin';

export const MyForm = () => (
-    <FormWithRedirect
+    <Form
        render={() => ...}
    />
);
```

> ### `<FormWithRedirect>` Props Have Changed

Reviewer: ```suggestion
### Form Props Have Changed
```

> `<FormWithRedirect>` used to accept [`react-final-form` `<Form>` props](https://final-form.org/docs/react-final-form/types/FormProps). It now accepts [`react-hook-form` `useForm` props](https://react-hook-form.com/api/useform). This also affects the other form components (`SimpleForm`, `TabbedForm`, etc.)

Reviewer: ```suggestion
`<FormWithRedirect>` used to accept [`react-final-form` `<Form>` props](https://final-form.org/docs/react-final-form/types/FormProps). It now accepts [`react-hook-form` `useForm` props](https://react-hook-form.com/api/useform). This also affects the other form components (`<SimpleForm>`, `<TabbedForm>`, etc.)
```

> The most commonly used prop is probably `initialValues` which is now named `defaultValues`:

Reviewer: ```suggestion
The most commonly used prop is probably `initialValues`, which is now named `defaultValues`:
```

```diff
const PostCreate = () => (
    <Create>
        <SimpleForm
-            initialValues={{ title: 'A default title' }}
+            defaultValues={{ title: 'A default title' }}
        >
            ...
        </SimpleForm>
    </Create>
)
```

> However, we kept the `validate` function prop which we automatically translate as a custom [`react-hook-form` `resolver`](https://react-hook-form.com/api/useform#validationResolver).

Reviewer: ```suggestion
We kept the `validate` function prop, which we automatically translate to a custom [`react-hook-form` `resolver`](https://react-hook-form.com/api/useform#validationResolver). So even if it's not technically a react-hook-form prop, you can still use `validate` as before.
```

> This also means you can now use [`yup`](https://github.com/jquense/yup), [`zod`](https://github.com/colinhacks/zod), [`joi`](https://github.com/sideway/joi), [superstruct](https://github.com/ianstormtaylor/superstruct), [vest](https://github.com/ealush/vest) or any [resolvers](https://react-hook-form.com/api/useform#validationResolver) supported by `react-hook-form` to apply schema validation.

Reviewer: ```suggestion
This also means you can now use [`yup`](https://github.com/jquense/yup), [`zod`](https://github.com/colinhacks/zod), [`joi`](https://github.com/sideway/joi), [superstruct](https://github.com/ianstormtaylor/superstruct), [vest](https://github.com/ealush/vest) or any [resolver](https://react-hook-form.com/api/useform#validationResolver) supported by `react-hook-form` to apply schema validation.
```

> ### `<FormWithRedirect>` Render Function Arguments Have Changed

Reviewer: I thought we got rid of the render prop?

Author: Not yet and I'm not sure we actually can

> The `<FormWithRedirect>` render function used to be called with an object containing parts of the `final-form` form state (`valid`, `invalid`, `pristine`, `dirty`). It now only receive the `handleSubmit` function which must be passed to the `onSubmit` prop of the underlying form. If you need access to the form state, call [`react-hook-form` `useFormState` hook](https://react-hook-form.com/api/useformstate):

Reviewer: ```suggestion
`<FormWithRedirect>` used to call its child function with an object containing parts of the `final-form` form state (`valid`, `invalid`, `pristine`, `dirty`). It now only passes the `handleSubmit` function, which must be passed down to the `onSubmit` prop of the underlying form. If you need to access the form state, call [the react-hook-form `useFormState` hook](https://react-hook-form.com/api/useformstate):
```

```diff
import { FormWithRedirect } from 'react-admin';
+ import { useFormState } from 'react-hook-form';

const MyCustomForm = () => {
    return (
-        <FormWithRedirect
+        <Form
-            render={({ valid, dirty, handleSubmit }) => (
+            render={({ handleSubmit }) => (
                <form onSubmit={handleSubmit}>
                    ...
-                    <SubmitButton disabled={!dirty || !valid}>Save</SubmitButton>
+                    <SubmitButton>Save</SubmitButton>
                </form>
            )}
        />
    );
};

-const SubmitButton = ({ disabled, ...props }) => {
+const SubmitButton = (props) => {
+    const { isDirty, isValid } = useFormState();
    return (
-        <button disabled={disabled} {...props} />
+        <button disabled={!isDirty || !isValid} {...props} />
    );
}
```

> ### `useInput` Signature Changed

Reviewer: You're listing changes to all Input components. Maybe the title of the section should be changed.

> `useInput` used to accept both a `initialValue` and a `defaultValue` prop. It now only accept `defaultValue`:

Reviewer: ```suggestion
`useInput` used to accept both a `initialValue` and a `defaultValue` prop. It now only accepts `defaultValue`:
```

```diff
const PostCreate = () => (
    <Create>
        <SimpleForm>
-            <TextInput source="title" initialValue="A default">
+            <TextInput source="title" defaultValue="A default">
        </SimpleForm>
    </Create>
)
```

`useInput` used to return `final-form` properties such as `input ` and `meta`. It now returns `field`, `fieldState` and `formState` (see https://react-hook-form.com/api/usecontroller).

```diff
import TextField from '@material-ui/core/TextField';
import { useInput, required } from 'react-admin';

const MyInput = ({ helperText, ...props }) => {
    const {
-        input: { name, onChange, ...rest },
+        field,
-        meta: { touched, error },
+        fieldState: { isTouched, invalid, error },
+        formState: { isSubmitted }
        isRequired
    } = useInput(props);

    return (
        <TextField

>             {...field}

Reviewer: there should be a diff here

            label={props.label}
            error={(isTouched || isSubmitted) && invalid}
            helperText={(isTouched || isSubmitted) && invalid ? error : helperText}
            required={isRequired}
            {...rest}
        />
    );
};
```

## `allowEmpty` Has Been Removed From `SelectInput`, `AutocompleteInput` and `AutocompleteArrayInput`

> This is necessary for the underlying MaterialUI inputs as they require that the current input value is actually one of the available option. Those components now always accept an empty value (an empty string by default). If you require the input to have a non empty value, use the `required` validation. You can safely remove this prop.

Reviewer 2: ```suggestion
This is necessary for the underlying MaterialUI inputs as they require that the current input value is actually one of the available option. Those components now always accept an empty value (an empty string by default). If you require the input to have a non-empty value, use the `required` validation. You can safely remove this prop.
```

> This is necessary for the underlying MaterialUI inputs as they require that the current input value is actually one of the available option. Those components now always accept an empty value (an empty string by default). If you require the input to have a non-empty value, use the `required` validation. You can safely remove this prop.

Reviewer: Not super clear, please elaborate and add a diff.
